diff -urp origin/iwl3945-base.c new/iwl3945-base.c
--- origin/iwl3945-base.c	2007-09-03 14:30:59.000000000 +0800
+++ new/iwl3945-base.c	2007-09-04 11:08:04.000000000 +0800
@@ -4851,7 +4851,11 @@ static void iwl_irq_tasklet(struct iwl_p
 	spin_unlock_irqrestore(&priv->lock, flags);
 }
 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2,6,19)
+static irqreturn_t iwl_isr(int irq, void *data, struct pt_regs *regs)
+#else
 static irqreturn_t iwl_isr(int irq, void *data)
+#endif
 {
 	struct iwl_priv *priv = data;
 	u32 inta, inta_mask;
diff -urp origin/iwl4965-base.c new/iwl4965-base.c
--- origin/iwl4965-base.c	2007-09-03 14:32:35.000000000 +0800
+++ new/iwl4965-base.c	2007-09-04 11:08:04.000000000 +0800
@@ -5134,7 +5134,11 @@ static void iwl_irq_tasklet(struct iwl_p
 	spin_unlock_irqrestore(&priv->lock, flags);
 }
 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2,6,19)
+static irqreturn_t iwl_isr(int irq, void *data, struct pt_regs *regs)
+#else
 static irqreturn_t iwl_isr(int irq, void *data)
+#endif
 {
 	struct iwl_priv *priv = data;
 	u32 inta, inta_mask;
