diff -urp old/iwl-helpers.h origin/iwl-helpers.h
--- old/iwl-helpers.h	2007-10-30 16:05:30.000000000 +0800
+++ origin/iwl-helpers.h	2007-10-30 16:25:28.000000000 +0800
@@ -150,6 +150,11 @@ static inline struct ieee80211_conf *iee
 #define IEEE80211_STYPE_BACK_REQ	0x0080
 #define IEEE80211_STYPE_BACK		0x0090
 
+/* Backward compat added via patches/06-csa.patch */
+#ifndef IEEE80211_CONF_CHANNEL_SWITCH
+#define IEEE80211_CONF_CHANNEL_SWITCH (1<<3)
+#endif
+
 
 static inline int ieee80211_is_management(u16 fc)
 {
