diff -urp origin/iwl3945-base.c new/iwl3945-base.c
--- origin/iwl3945-base.c	2007-09-03 14:30:59.000000000 +0800
+++ new/iwl3945-base.c	2007-09-04 11:27:24.000000000 +0800
@@ -8348,6 +8348,7 @@ static int iwl_pci_probe(struct pci_dev 
 	    IEEE80211_HW_HOST_GEN_BEACON_TEMPLATE;
 
 	hw->queues = 4;
+	hw->preferred_rate_control = NULL;
 
 	spin_lock_init(&priv->lock);
 	spin_lock_init(&priv->power_data.lock);
diff -urp origin/iwl-3945-rs.c new/iwl-3945-rs.c
--- origin/iwl-3945-rs.c	2007-09-03 14:30:59.000000000 +0800
+++ new/iwl-3945-rs.c	2007-09-04 11:27:24.000000000 +0800
@@ -957,11 +957,13 @@ void iwl_rate_scale_init(struct ieee8021
 
 void iwl3945_rate_control_register(struct ieee80211_hw *hw)
 {
+	hw->preferred_rate_control = (char *)rs_ops.name;
 	ieee80211_rate_control_register(&rs_ops);
 }
 
 void iwl3945_rate_control_unregister(struct ieee80211_hw *hw)
 {
+	hw->preferred_rate_control = NULL;
 	ieee80211_rate_control_unregister(&rs_ops);
 }
 
diff -urp origin/iwl4965-base.c new/iwl4965-base.c
--- origin/iwl4965-base.c	2007-09-03 14:32:35.000000000 +0800
+++ new/iwl4965-base.c	2007-09-04 11:27:24.000000000 +0800
@@ -8920,6 +8920,7 @@ static int iwl_pci_probe(struct pci_dev 
 	    IEEE80211_HW_HOST_GEN_BEACON_TEMPLATE;
 
 	hw->queues = 4;
+	hw->preferred_rate_control = NULL;
 #ifdef CONFIG_IWL4965_HT
 #ifdef CONFIG_IWL4965_HT_AGG
 	hw->queues = 16;
diff -urp origin/iwl-4965-rs.c new/iwl-4965-rs.c
--- origin/iwl-4965-rs.c	2007-09-03 14:30:59.000000000 +0800
+++ new/iwl-4965-rs.c	2007-09-04 11:27:24.000000000 +0800
@@ -2105,11 +2105,13 @@ void iwl_rate_scale_init(struct ieee8021
 
 void iwl4965_rate_control_register(struct ieee80211_hw *hw)
 {
+	hw->preferred_rate_control = (char *)rs_ops.name;
 	ieee80211_rate_control_register(&rs_ops);
 }
 
 void iwl4965_rate_control_unregister(struct ieee80211_hw *hw)
 {
+	hw->preferred_rate_control = NULL;
 	ieee80211_rate_control_unregister(&rs_ops);
 }
 
