diff -urp origin/iwl-3945-rs.c new/iwl-3945-rs.c
--- origin/iwl-3945-rs.c	2007-08-23 11:28:13.000000000 +0800
+++ new/iwl-3945-rs.c	2007-08-23 11:35:43.000000000 +0800
@@ -973,11 +973,13 @@ void iwl_rate_scale_init(struct ieee8021
 
 void iwl_rate_control_register(struct ieee80211_hw *hw)
 {
+	hw->preferred_rate_control = (char *)rs_ops.name;
 	ieee80211_rate_control_register(&rs_ops);
 }
 
 void iwl_rate_control_unregister(struct ieee80211_hw *hw)
 {
+	hw->preferred_rate_control = NULL;
 	ieee80211_rate_control_unregister(&rs_ops);
 }
 
Only in new: iwl-3945-rs.c.orig
diff -urp origin/iwl-4965-rs.c new/iwl-4965-rs.c
--- origin/iwl-4965-rs.c	2007-08-23 11:28:52.000000000 +0800
+++ new/iwl-4965-rs.c	2007-08-23 11:35:43.000000000 +0800
@@ -2150,11 +2150,13 @@ void iwl_rate_scale_init(struct ieee8021
 
 void iwl_rate_control_register(struct ieee80211_hw *hw)
 {
+	hw->preferred_rate_control = (char *)rs_ops.name;
 	ieee80211_rate_control_register(&rs_ops);
 }
 
 void iwl_rate_control_unregister(struct ieee80211_hw *hw)
 {
+	hw->preferred_rate_control = NULL;
 	ieee80211_rate_control_unregister(&rs_ops);
 }
 
Only in new: iwl-4965-rs.c.orig
diff -urp origin/iwl-base.c new/iwl-base.c
--- origin/iwl-base.c	2007-08-23 10:49:44.000000000 +0800
+++ new/iwl-base.c	2007-08-23 11:35:43.000000000 +0800
@@ -9209,6 +9209,7 @@ static int iwl_pci_probe(struct pci_dev 
 	    IEEE80211_HW_HOST_GEN_BEACON_TEMPLATE;
 
 	hw->queues = 4;
+	hw->preferred_rate_control = NULL;
 #if IWL == 4965
 #ifdef CONFIG_IWLWIFI_HT
 #ifdef CONFIG_IWLWIFI_HT_AGG
Only in new: iwl-base.c.orig
