From 2daad64f3facaaef4f13b8bd514f82b44a4fc030 Mon Sep 17 00:00:00 2001
From: Kernel PPA uploader dude <kernel-ppa@zinc.canonical.com>
Date: Wed, 25 Feb 2009 16:42:55 +0000
Subject: [PATCH] mainline build: allow us to force things on in the config

Signed-off-by: Andy Whitcroft <apw@canonical.com>
---
 debian.master/scripts/misc/oldconfig |    7 ++++++-
 1 files changed, 6 insertions(+), 1 deletions(-)

--- a/debian.master/scripts/misc/oldconfig.orig	2009-05-14 12:52:21.000000000 +0100
+++ b/debian.master/scripts/misc/oldconfig	2009-05-14 12:53:32.000000000 +0100
@@ -55,7 +55,13 @@
 for config in $configs; do
 	echo "Running $what for $config ... "
 
-	cat $confdir/$config > .config
+	if [ -f "$BUILD_CONFIG_OVERRIDE" ]; then
+		echo "Adding $BUILD_CONFIG_OVERRIDE to $config ... "
+		cat "$BUILD_CONFIG_OVERRIDE" $confdir/$config \
+			"$BUILD_CONFIG_OVERRIDE">.config
+	else
+		cat $confdir/$config > .config
+	fi
 
 	make -C ../ O=`pwd` "$what" ARCH=$kernarch
 
