Index: compatible/net/mac80211/ieee80211_sta.c
===================================================================
--- compatible.orig/net/mac80211/ieee80211_sta.c	2008-01-06 13:56:59.000000000 +0200
+++ compatible/net/mac80211/ieee80211_sta.c	2008-01-06 13:57:56.000000000 +0200
@@ -2521,9 +2521,11 @@
 	struct ieee80211_local *local = wdev_priv(dev->ieee80211_ptr);
 	struct sk_buff *skb;
 	struct ieee80211_mgmt *mgmt;
-	u16 capab=0;
+	u16 capab;
 
-	skb = dev_alloc_skb(sizeof(*mgmt) + local->hw.extra_tx_headroom);
+	skb = dev_alloc_skb(sizeof(*mgmt) + local->hw.extra_tx_headroom +
+			    sizeof(mgmt->u.action.category) +
+			    sizeof(mgmt->u.action.u.addba_resp));
 	if (!skb) {
 		printk(KERN_DEBUG "%s: failed to allocate buffer "
 		       "for addba resp frame\n", dev->name);
@@ -2542,12 +2544,13 @@
 	mgmt->frame_control = IEEE80211_FC(IEEE80211_FTYPE_MGMT,
 					   IEEE80211_STYPE_ACTION);
 
-	skb_put(skb, 1 + sizeof(mgmt->u.action.u.addba_resp));
+	skb_put(skb, sizeof(mgmt->u.action.category) +
+		sizeof(mgmt->u.action.u.addba_resp));
 	mgmt->u.action.category = WLAN_CATEGORY_BACK;
 	mgmt->u.action.u.addba_resp.action_code = WLAN_ACTION_ADDBA_RESP;
 	mgmt->u.action.u.addba_resp.dialog_token = dialog_token;
 
-	capab |= (u16)(policy << 1);		/* bit 1 aggregation policy (1 - immediate 0 - delayed )*/
+	capab = (u16)(policy << 1);		/* bit 1 aggregation policy (1 - immediate 0 - delayed )*/
 	capab |= (u16)(tid << 2); 			/* bit 5:2 TID number */
 	capab |= (u16)(buf_size << 6);		/* bit 15:6 max size of aggergation */
 
@@ -2744,11 +2747,11 @@
 	struct ieee80211_if_sta *ifsta = &sdata->u.sta;
 	struct sk_buff *skb;
 	struct ieee80211_mgmt *mgmt;
-	u16 capab=0;
-
-	skb = dev_alloc_skb(sizeof(*mgmt) + local->hw.extra_tx_headroom + 1 +
-				sizeof(mgmt->u.action.u.addba_req));
+	u16 capab;
 
+	skb = dev_alloc_skb(sizeof(*mgmt) + local->hw.extra_tx_headroom +
+			    sizeof(mgmt->u.action.category) +
+			    sizeof(mgmt->u.action.u.delba));
 
 	if (!skb) {
 		printk(KERN_ERR "%s: failed to allocate buffer "
@@ -2768,7 +2771,8 @@
 	mgmt->frame_control = IEEE80211_FC(IEEE80211_FTYPE_MGMT,
 					IEEE80211_STYPE_ACTION);
 
-	skb_put(skb, 1 + sizeof(mgmt->u.action.u.addba_req));
+	skb_put(skb, sizeof(mgmt->u.action.category) +
+		sizeof(mgmt->u.action.u.addba_req));
 
 	mgmt->u.action.category = WLAN_CATEGORY_BACK;
 	mgmt->u.action.u.addba_req.action_code = WLAN_ACTION_ADDBA_REQ;
@@ -2853,8 +2857,9 @@
 	struct ieee80211_mgmt *mgmt;
 	u16 params=0;
 
-	skb = dev_alloc_skb(sizeof(*mgmt) + local->hw.extra_tx_headroom + 1 +
-							sizeof(mgmt->u.action.u.delba));
+	skb = dev_alloc_skb(sizeof(*mgmt) + local->hw.extra_tx_headroom +
+			    sizeof(mgmt->u.action.category) +
+			    sizeof(mgmt->u.action.u.delba));
 
 	if (!skb) {
 		printk(KERN_ERR "%s: failed to allocate buffer "
@@ -2873,7 +2878,8 @@
 		memcpy(mgmt->bssid, ifsta->bssid, ETH_ALEN);
 	mgmt->frame_control = IEEE80211_FC(IEEE80211_FTYPE_MGMT,IEEE80211_STYPE_ACTION);
 
-	skb_put(skb, 1 + sizeof(mgmt->u.action.u.delba));
+	skb_put(skb, sizeof(mgmt->u.action.category) +
+		sizeof(mgmt->u.action.u.delba));
 
 	mgmt->u.action.category = WLAN_CATEGORY_BACK;
 	mgmt->u.action.u.delba.action_code = WLAN_ACTION_DELBA;
