diff --git a/net/mac80211/ieee80211_ioctl.c b/net/mac80211/ieee80211_ioctl.c
index e7904db..3320f12 100644
--- a/net/mac80211/ieee80211_ioctl.c
+++ b/net/mac80211/ieee80211_ioctl.c
@@ -693,6 +693,7 @@ static int ieee80211_ioctl_siwscan(struc
 	struct ieee80211_sub_if_data *sdata = IEEE80211_DEV_TO_SUB_IF(dev);
 	u8 *ssid = NULL;
 	size_t ssid_len = 0;
+	struct iw_scan_req* scan = (struct iw_scan_req*) extra;
 
 	if (!netif_running(dev))
 		return -ENETDOWN;
@@ -700,9 +701,16 @@ static int ieee80211_ioctl_siwscan(struc
 	switch (sdata->type) {
 	case IEEE80211_IF_TYPE_STA:
 	case IEEE80211_IF_TYPE_IBSS:
-		if (local->scan_flags & IEEE80211_SCAN_MATCH_SSID) {
+		if (data->flags & IW_SCAN_THIS_ESSID) {
+			ssid = scan->essid;
+			ssid_len = scan->essid_len;
+			local->scan_ssid_len = ssid_len;
+			memcpy(local->scan_ssid, ssid, ssid_len);
+		} else if (local->scan_flags & IEEE80211_SCAN_MATCH_SSID) {
 			ssid = sdata->u.sta.ssid;
 			ssid_len = sdata->u.sta.ssid_len;
+			local->scan_ssid_len = ssid_len;
+			memcpy(local->scan_ssid, ssid, ssid_len);
 		}
 		break;
 	case IEEE80211_IF_TYPE_AP:
diff --git a/net/mac80211/ieee80211_sta.c b/net/mac80211/ieee80211_sta.c
index 0d99b68..7c7be49 100644
--- a/net/mac80211/ieee80211_sta.c
+++ b/net/mac80211/ieee80211_sta.c
@@ -1978,7 +1978,10 @@ void ieee80211_sta_work(struct work_stru
 	if (ifsta->state != IEEE80211_AUTHENTICATE &&
 	    ifsta->state != IEEE80211_ASSOCIATE &&
 	    test_and_clear_bit(IEEE80211_STA_REQ_SCAN, &ifsta->request)) {
-		ieee80211_sta_start_scan(dev, NULL, 0);
+		if (local->scan_ssid_len)
+			ieee80211_sta_start_scan(dev, local->scan_ssid, local->scan_ssid_len);
+		else
+			ieee80211_sta_start_scan(dev, NULL, 0);
 		return;
 	}
 
